internal/pkg/app/app.go

```diff
@@ -0,0 +1,78 @@
+/**
+ *  Copyright (c) 2020  Comcast Cable Communications Management, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package app
+
+import (
+	"context"
+	"fmt"
+	"github.com/gorilla/mux"
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+	"go.uber.org/fx"
+	"net/http"
+)
+
+var Version = "v0.0.0"
+
+func NewLogger() *zerolog.Logger {
+	// Create a child logger off the global logger instead of creating a new logger from scratch because
+	// global loggers already have some context values that child logger can inherit
+	logger := log.Logger.With().Logger()
+	return &logger
+}
+
+func NewMux(r *mux.Router, middlewares []func(next http.Handler) http.Handler) (http.Handler, error) {
+	for _, middleware := range middlewares {
+		r.Use(middleware)
+	}
+	return r, nil
+}
+
+func SetupAPIServer(lifecycle fx.Lifecycle, config Config, logger *zerolog.Logger, mux http.Handler) error {
+	port := config.GetInt("api.port")
+	if port < 1 {
+		err := &InvalidOptionError{nil, fmt.Errorf("invalid port value %d", port)}
+		logger.Error().Msg(err.Error())
+		return err
+	}
+
+	server := &http.Server{
+		Addr:    ":" + config.GetString("api.port"),
+		Handler: mux,
+	}
+
+	lifecycle.Append(
+		fx.Hook{
+			OnStart: func(context.Context) error {
+				go server.ListenAndServe()
+				logger.Info().Msg("API Server Started")
+				return nil
+			},
+			OnStop: func(ctx context.Context) error {
+				err := server.Shutdown(ctx)
+				if err != nil {
```

Comment:
If you are handling the error by logging it, do not return an error.  If you return the error, you can always augment it with more information.

```
logger.Error().Str("op", "SetupAPIServer.OnStop").Msg(err.Error())
```
could become
```
return fmt.Errorf("could not stop API server: %w", err)
```

then you'd drop the `return err` at the end.

Reply:
So my preference is to log the error in OnStop so that I can have structured logs so that I can query the log more easily (ie, using op=SetupAPIServer.OnStop)

Reply:
@holidaymike Sure.  So log it as you prefer, and return `nil`. The code is still returning an error on stopping.

Let's not bring the bad habit of logging and returning errors into this project.  It should be one or the other.  This eliminates the issue that we currently have in our internal services where one error will end up producing 10 similar error log lines, which is:

* Expensive as far as data produced (storage + network)
* Confusing from a support standpoint when trying to figure out what is going on as no one can tell what the end result was (was it handled?)

Reply:
I am pretty sure that is what I've done (return nil).